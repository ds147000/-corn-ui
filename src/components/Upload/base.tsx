import React, { useEffect, useMemo, useRef } from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import Image from '../Image'
import previewImage from '../PreviewImage'
import Icon from '../Icon'

export declare namespace Upload {
  type Media = {
    mediaId: number
    content: string
  }

  /** 自定义上传，必须返回上传后的结果 */
  type handleUpload = (file: File) => Promise<Media>

  /** 上传完成回调函数 */
  type onChange = (imgs: Media[]) => void

  type layout = 'square' | 'row'

  type onRemove = (img: Media, index: number) => void

  interface props {
    /** 图片项类名 */
    itemClassName?: string
    /** 最多数量 */
    count?: number
    /** 布局类型 */
    layout?: layout
    /** 自定义渲染上传按钮 */
    btn?: React.ReactNode
    /** 已经上传文件对象 */
    list?: Media[]
    /** 上传项删除事件 */
    onRemove?: onRemove
  }
}

export const DEFAULT_COUNT = 9

export const UploadBase: React.FC<Upload.props> = ({
  itemClassName,
  count = DEFAULT_COUNT,
  layout = 'row',
  btn,
  list,
  onRemove
}) => {
  const El = useRef<HTMLDivElement>()
  const isShowBtn = !list || list.length < count

  const _class = useMemo(() => classNames(
    'xrk-upload',
    'xrk-upload-' + layout
  ), [ layout ])

  const _itemClass = useMemo(() => classNames(
    'xrk-upload-item',
    itemClassName
  ), [ itemClassName ])

  const onPreview = (index: number): void => {
    const _list = list as Upload.Media[]
    previewImage({
      current: _list[index].content,
      urls: _list?.map((item) => item.content)
    })
  }

  useEffect(() => {
    if (layout !== 'row') return

    // #if _APP === 'weapp'

    // #else
    try {
      El.current?.scrollTo?.({ left: El.current.scrollWidth, top: 0, behavior: 'smooth' })
    } catch (error) {
      El.current?.scrollTo?.(El.current.scrollWidth, 0)
    }
    // #endif
  }, [ list?.length, layout ])

  return (
    <View className={_class} ref={El} >
      {list?.map((item, key) => (
        <View key={item.mediaId + item.content} className={_itemClass} >
          <View
            data-testid="upload-item"
            className="xrk-upload-remove xrk-f xrk-ac xrk-jc"
            onClick={(): void => onRemove?.(item, key)}
          >
            <Icon name="delete" data-testid="upload-remove" />
          </View>
          <Image
            className="xrk-upload-img"
            src={item.content}
            onClick={(): void => onPreview(key)}
            lazyLoad
            mode="aspectFill"
          />
        </View>
      ))}
      {isShowBtn && btn}
    </View>
  )
}

