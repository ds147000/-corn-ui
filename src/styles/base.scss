@import '~@tarojs/components-react/src/components/view/style/index.css';
@import '~@tarojs/components-react/src/components/text/style/index.css';
@import '~@tarojs/components-react/src/components/image/style/index.css'
