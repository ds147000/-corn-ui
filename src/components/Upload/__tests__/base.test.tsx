import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { UploadBase, Upload } from '../base'
import { act } from 'react-dom/test-utils'

jest.mock('../../PreviewImage')

describe('UploadBase', () => {
  test('row', async () => {
    const screen = render(<UploadBase />)

    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('square', async () => {
    const screen = render(<UploadBase />)

    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })


  test('自定义按钮', async () => {
    const onClick = jest.fn()
    const screen = render(<UploadBase btn={<div onClick={onClick} >上传</div>} />)
    fireEvent.click(screen.getByText('上传'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
    await waitFor(() => expect(onClick).toHaveBeenCalledTimes(1))
  })

  test('list', async () => {
    const list: Upload.Media[] = [
      {
        mediaId: 100,
        content: '100.jpg',
        status: 'done'
      },
      {
        mediaId: 101,
        content: '101.jpg',
        status: 'done'
      },
      {
        mediaId: 102,
        content: '102.jpg',
        status: 'done'
      }
    ]

    const onClick = jest.fn()
    const screen = render(<UploadBase list={list} btn={<div onClick={onClick} >上传</div>} />)
    fireEvent.click(screen.getByText('上传'))
    await waitFor(() => expect(onClick).toHaveBeenCalledTimes(1))
    await waitFor(async () => expect(await screen.queryAllByTestId('upload-item').length).toBe(3))
  })

  test('list and itemClass', async () => {
    const list: Upload.Media[] = [
      {
        mediaId: 100,
        content: '100.jpg',
        status: 'done'
      },
      {
        mediaId: 101,
        content: '101.jpg',
        status: 'done'
      },
      {
        mediaId: 102,
        content: '102.jpg',
        status: 'done'
      }
    ]

    const onClick = jest.fn()
    const screen = render(<UploadBase itemClassName="abs100" list={list} btn={<div onClick={onClick} >上传</div>} />)
    fireEvent.click(screen.getByText('上传'))
    await waitFor(() => expect(onClick).toHaveBeenCalledTimes(1))
    await waitFor(async () => expect(await screen.queryAllByTestId('upload-item').length).toBe(3))
    await waitFor(async () => expect(screen.container.querySelectorAll('.abs100').length).toBe(3))
  })

  test('list and count', async () => {
    const list: Upload.Media[] = [
      {
        mediaId: 100,
        content: '100.jpg',
        status: 'done'
      },
      {
        mediaId: 101,
        content: '101.jpg',
        status: 'done'
      },
      {
        mediaId: 102,
        content: '102.jpg',
        status: 'done'
      }
    ]

    const screen = render(<UploadBase count={3} list={list} btn={<div>上传</div>} />)
    await waitFor(() => expect(screen.queryByTestId('上传')).toBeNull())
    await waitFor(async () => expect(await screen.queryAllByTestId('upload-item').length).toBe(3))
  })

  test('list and remove', async () => {
    const onRemove = jest.fn()
    const list: Upload.Media[] = [
      {
        mediaId: 100,
        content: '100.jpg',
        status: 'done'
      },
      {
        mediaId: 101,
        content: '101.jpg',
        status: 'done'
      },
      {
        mediaId: 102,
        content: '102.jpg',
        status: 'done'
      }
    ]

    const screen = render(<UploadBase layout="square" onRemove={onRemove} list={list} btn={<div>上传</div>} />)
    const items = await screen.queryAllByTestId('upload-remove')
    fireEvent.click(items[0])
    expect(onRemove).toHaveBeenCalledTimes(1)
    expect(onRemove.mock.calls[0][0]).toEqual(list[0])
  })

  test('onPreview', () => {
    const list: Upload.Media[] = [
      {
        mediaId: 100,
        content: '100.jpg',
        status: 'done'
      },
      {
        mediaId: 101,
        content: '101.jpg',
        status: 'loading'
      },
      {
        mediaId: 102,
        content: '102.jpg',
        status: 'error'
      }
    ]

    act(() => {
      const screen = render(<UploadBase list={list} btn={<div>上传</div>} />)
      fireEvent.click(screen.container.querySelector('.corn-upload-img'))
    })
  })

})
