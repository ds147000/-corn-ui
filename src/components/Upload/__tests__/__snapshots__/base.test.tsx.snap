// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadBase row 1`] = `
<div>
  <div
    class="xrk-upload xrk-upload-row"
  />
</div>
`;

exports[`UploadBase square 1`] = `
<div>
  <div
    class="xrk-upload xrk-upload-row"
  />
</div>
`;

exports[`UploadBase 自定义按钮 1`] = `
<div>
  <div
    class="xrk-upload xrk-upload-row"
  >
    <div>
      上传
    </div>
  </div>
</div>
`;
